### ISSUE TITLE
Improve JWT usability for getting started experience

### ISSUE THREAD

**Issue by emmanuelbernard** [quarkusio/quarkus#5211]
Just by adding jwt to the extension and nothing else leads to the following error

```
2019-11-05 11:32:49,285 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-3) HTTP Request to /hello failed, error id: 892c74b3-bf30-4e11-bba9-cc24d5c61cf5-3: java.lang.IllegalStateException: JWTAuthContextInfo has not been initialized. Please make sure that either 'mp.jwt.verify.publickey' or 'mp.jwt.verify.publickey.location' properties are set.
        at io.smallrye.jwt.config.JWTAuthContextInfoProvider.lambda$throwException$0(JWTAuthContextInfoProvider.java:382)
        at java.util.Optional.orElseThrow(Optional.java:290)
        at io.smallrye.jwt.config.JWTAuthContextInfoProvider.getContextInfo(JWTAuthContextInfoProvider.java:376)
        at io.smallrye.jwt.config.JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_Bean.create(JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_Bean.zig:132)
        at io.smallrye.jwt.config.JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_Bean.create(JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_Bean.zig:184)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:72)
        at io.quarkus.arc.impl.ComputingCache$CacheFunction.lambda$apply$0(ComputingCache.java:99)
        at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
        at io.quarkus.arc.impl.ComputingCache.getValue(ComputingCache.java:41)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:23)
        at io.smallrye.jwt.config.JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_ClientProxy.arc$delegate(JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_ClientProxy.zig:1026)
        at io.smallrye.jwt.config.JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_ClientProxy.getTokenHeader(JWTAuthContextInfoProvider_ProducerMethod_getContextInfo_21d111677ae04ef1a7cf911100af3482e4c3b30a_ClientProxy.zig:474)
        at io.smallrye.jwt.auth.AbstractBearerTokenExtractor.getBearerToken(AbstractBearerTokenExtractor.java:56)
        at io.quarkus.smallrye.jwt.runtime.auth.JWTAuthMechanism.authenticate(JWTAuthMechanism.java:47)
        at io.quarkus.smallrye.jwt.runtime.auth.JWTAuthMechanism_ClientProxy.authenticate(JWTAuthMechanism_ClientProxy.zig:72)
        at io.quarkus.vertx.http.runtime.security.HttpAuthenticator.attemptAuthentication(HttpAuthenticator.java:66)
        at io.quarkus.vertx.http.runtime.security.HttpAuthenticator_ClientProxy.attemptAuthentication(HttpAuthenticator_ClientProxy.zig:150)
        at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$1.handle(HttpSecurityRecorder.java:35)
        at io.quarkus.vertx.http.runtime.security.HttpSecurityRecorder$1.handle(HttpSecurityRecorder.java:24)
        at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1034)
        at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:131)
        at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:130)
        at io.quarkus.vertx.http.deployment.devmode.VertxHotReplacementSetup$2.handle(VertxHotReplacementSetup.java:75)
        at io.quarkus.vertx.http.deployment.devmode.VertxHotReplacementSetup$2.handle(VertxHotReplacementSetup.java:64)
        at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:330)
        at io.vertx.core.impl.ContextImpl.executeTask(ContextImpl.java:369)
        at io.vertx.core.impl.EventLoopContext.lambda$executeAsync$0(EventLoopContext.java:38)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:510)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:518)
        at io.netty.util.concurrent.SingleThreadEventExecutor$6.run(SingleThreadEventExecutor.java:1044)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)

```

Are these absolutly mandatory and could we not raise an error if JWT restrictions are not used in the app anywhere?

CC @kenfinnigan

**Comment by sberyozkin** [comment]
@emmanuelbernard 
One can use `smallrye-jwt` simply to verify the tokens and make sure the roles info extracted from them is effective against `@RolesAllowed` restrictions - so there is no requirement for the users to add MP JWT `JsonWebToken` or `Claim` in their code.
The extension can be 'disabled' with a property, which is a common pattern in Quarkus AFAIK.

**Comment by kenfinnigan** [comment]
I think @emmanuelbernard is asking for an error to be displayed at build time if the extension is present but the required configuration isn't set, as opposed to failing to respond to an endpoint at runtime

**Comment by benneq** [comment]
I have a similar issue, but with a completely different use case:

My Quarkus app is running behind an API Gateway. The API Gateway is responsible for validating the JWT. Though it's unnecessary to validate the JWT again within the Quarkus app. **But** I still want to be able to access the JWT's payload.

The request flow works like this:
1. Client sends request to `https://example.org/foo/bar`
2. The API Gateway validates the JWT from the `Authorization` Header
3. if (valid) : Pass the request to matching Quarkus app, that is responsible for handling `/foo/bar`

In other words: I'd like to use `@Inject JsonWebToken jwt;` and `@Inject @Claim`, but the app should **only** decode the JWT and **not** verify it's signature.

**Comment by chrisgleissner** [comment]
I just created a Quarkus project from https://code.quarkus.io/ and selected JWT as one of the options. Building the resulting project without further modifications, I get the error described above and exactly 4 Google results when I search for a way how to resolve it.

I suggest there should either be a much clearer error message or - even better - there should not be  an error at all. Instead, JWT tokens should be disabled by default, even when selecting this feature to be part of the project, and the user should have to explicitly enable this feature via the Quarkus properties. 

It is a suboptimal first Quarkus experience if the project generated by Quarkus' own starter fails.

**Comment by sberyozkin** [comment]
@benneq We are getting closer to supporting the custom verification factories thanks to @andreas-eberle so if the users will have an option eventually to plugin the alternative mechanism to get `JsonWebToken` out of the header . The verification is an essential part of smallrye-jwt, providing a 1st class support for disabling the verification with a property will just get us a stream of CVEs. Lets disicuss it in the smallrye-jwt issue.

**Comment by benneq** [comment]
@sberyozkin Could you link the smallrye-jwt issue for custom verifications factories?
